Implementation of Weighted CRF Tagger (handling unbalanced datasets)

eraldoluis opened the pull request:
<!-- To ensure we can review your pull request promptly please complete this template entirely. -->

<!-- Please reference the issue number here. You can replace "Fixes" with "Closes" if it makes more sense. -->
Closes #4619 .

Dependency of [allennlp-models PR #341](https://github.com/allenai/allennlp-models/pull/341)

Changes proposed in this pull request:
<!-- Please list all changes/additions here. -->

- I implemented and [experimentally compared](https://eraldoluis.github.io/2022/05/10/weighted-crf.html) three sample weighting strategies for CrfTagger.
- The three strategies are implemented in the files `allennlp/modules/conditional_random_field_**<strategy>**.py`, where **<strategy>** can be: `wemission`, `wtrans`, or `lannoy`.
- I generalized the methods `_input_likelihood(...)` and `_joint_likelihood(...)` of the `ConditionalRandomField` class so that they now receive an argument with the transition weights. In that way, I could implement the two basic sample weighting strategies (wemission and wtrans) by just subclassing this class and weighting the corresponding weights (`logits` and `transitions`) in the `forward(...)` method before calling `_input_likelihood(...)` and `_joint_likelihood(...)`. No modification was necessary to the basic algorithms in these two methods.
- On the other hand, the strategy proposed by [Lannoy et al.](https://perso.uclouvain.be/michel.verleysen/papers/ieeetbe12gdl.pdf) (suggestion by @dirkgr) needed a quite different implementation.

## Before submitting

<!-- Please complete this checklist BEFORE submitting your PR to speed along the review process. -->
- [x] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

<!-- Please complete this checklist AFTER submitting your PR to speed along the review process. -->
- [ ] All GitHub Actions jobs for my pull request have passed.
- [ ] **`codecov/patch`** reports high test coverage (at least 90%).
    You can find this under the "Actions" tab of the pull request once the other checks have finished.

epwalsh commented:
Hi @eraldoluis, thanks for this! I may not have time for a thorough review this week but this will be a priority next week.

eraldoluis commented:
> Hi @eraldoluis, thanks for this! I may not have time for a thorough review this week but this will be a priority next week.

Thank you, @epwalsh !

epwalsh commented on allennlp/modules/conditional_random_field_wemission.py:
```diff
@@ -0,0 +1,82 @@
+        if label_weights is None:
+            raise ConfigurationError("label_weights must be given")
+
+        self.label_weights = torch.nn.Parameter(torch.Tensor(label_weights), requires_grad=False)
```
I think it might be better to use [`self.register_buffer()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer) here instead of defining the weights as a parameter. That way we can be sure the label weights aren't passed to the optimizer.

https://discuss.pytorch.org/t/what-is-the-difference-between-register-buffer-and-register-parameter-of-nn-module/32723/11

eraldoluis replied:
Done!

epwalsh commented on allennlp/modules/conditional_random_field_wemission.py:
```diff
@@ -0,0 +1,82 @@
+
+class ConditionalRandomFieldWeightEmission(ConditionalRandomField):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
```
Could you add something about the weighting strategy here?

eraldoluis replied:
Done!

epwalsh commented on allennlp/modules/conditional_random_field_wtrans.py:
```diff
@@ -0,0 +1,87 @@
+
+class ConditionalRandomFieldWeightTrans(ConditionalRandomField):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
```
Should also have a note about the weighting strategy here.

eraldoluis replied:
Done!

epwalsh commented on allennlp/modules/conditional_random_field_lannoy.py:
```diff
@@ -0,0 +1,486 @@
+
+class ConditionalRandomFieldLannoy(torch.nn.Module):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
+
```
Also would be good to have a note about the weight strategy here.

eraldoluis replied:
Done!

epwalsh commented on allennlp/modules/conditional_random_field_lannoy.py:
```diff
@@ -0,0 +1,486 @@
+VITERBI_DECODING = Tuple[List[int], float]  # a list of tags, and a viterbi score
+
+
+def allowed_transitions(constraint_type: str, labels: Dict[int, str]) -> List[Tuple[int, int]]:
```
Is this any different from the same function in `conditional_random_field.py`?

eraldoluis replied:
I simply forgot to adapt this class. I think now it is much better.

epwalsh commented on allennlp/modules/conditional_random_field_lannoy.py:
```diff
@@ -0,0 +1,486 @@
+    return allowed
+
+
+def is_transition_allowed(
```
Same question here: is any different from the original?

eraldoluis replied:
Idem.

epwalsh commented on allennlp/modules/conditional_random_field_lannoy.py:
```diff
@@ -0,0 +1,486 @@
+            for i, j in constraints:
+                constraint_mask[i, j] = 1.0
+
+        self._constraint_mask = torch.nn.Parameter(constraint_mask, requires_grad=False)
```
This should probably be a buffer as well, but I just realized this is how it's done in the original CRF module, so I guess I'm okay with this.

eraldoluis replied:
Ok. I didn't touch it. Although I agree with you that this should be a buffer as well.

epwalsh reviewed (commented):
Hey @eraldoluis, this looks really great! This is my first pass through (I still need to go through some of the code in more detail to get my head around it), but I just have some minor comments. In addition to the comments below, I will also say I think we should organize these modules into a common parent module. That is, create a new folder `allennlp/modules/conditional_random_field/` and move the 3 implementations into there.

eraldoluis commented:
Thank you a lot @epwalsh for the effort you put on this.

I tried to address your first concerns. Let me know what you think about my changes.

I am looking forward to your feedback regarding the whole thing. Let me know if you have any questions. I will be happy to discuss this further if necassary.

epwalsh commented:
Thanks for the quick responses/fixes! Changes look good. I should clarify what I meant by:

> create a new folder allennlp/modules/conditional_random_field/ and move the 3 implementations into there.

Looks like you left `allennlp/modules/conditional_random_field.py` where it is, and then moved the weighted CRFs into `allennlp/modules/conditional_random_field_weighted/`. I'd rather have a single submodule (folder) called `allennlp/modules/conditional_random_field/` with all of the CRFs (included the non-weighted base class).

epwalsh commented:
I liked your blog post a lot by the way!

eraldoluis commented:
> Looks like you left `allennlp/modules/conditional_random_field.py` where it is, and then moved the weighted CRFs into `allennlp/modules/conditional_random_field_weighted/`. I'd rather have a single submodule (folder) called `allennlp/modules/conditional_random_field/` with all of the CRFs (included the non-weighted base class).

Yes. I was unsure at first. But now I renamed the module to `conditional_random_field` and moved the original class to it. I also updated the changelog, which I had forgotten.

I also updated `allennlp-models` to reflect the new module organization. Unfortunately, I pushed first to the allennlp repository and the Model Tests failed (because `allennlp-models` was outdated). But these tests should pass now.

Let me know what do you think.

eraldoluis commented:
Thank you very much, @epwalsh and @dirkgr ! This was my first contribution for an open source project and it was quite fun. I will definitely try it again soon. :)
